atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLogEvents.java

```diff
@@ -94,10 +99,23 @@ default CommitUpdate toCommitUpdate(
         verifyReturnedEventsEnclosesTransactionVersions(startVersion.version() + 1, endVersion.version());
 
         LockEventVisitor eventVisitor = new LockEventVisitor(commitInfo.map(CommitInfo::commitLockToken));
-        Set<LockDescriptor> locksTakenOut = events().events().stream()
-                .flatMap(event -> event.accept(eventVisitor).stream())
+
+        List<LocksAndMetadata> locksAndMetadata = events().events().stream()
+                .map(event -> event.accept(eventVisitor))
+                .collect(Collectors.toUnmodifiableList());
+        Set<LockDescriptor> locksTakenOut = locksAndMetadata.stream()
+                .map(LocksAndMetadata::lockDescriptors)
+                .flatMap(Set::stream)
                 .collect(ImmutableSet.toImmutableSet());
-        return CommitUpdate.invalidateSome(locksTakenOut);
+        Map<LockDescriptor, List<ChangeMetadata>> locksWithAggregatedMetadata = locksAndMetadata.stream()
+                .map(LocksAndMetadata::metadata)
+                .flatMap(Optional::stream)
+                .map(LockRequestMetadata::lockDescriptorToChangeMetadata)
+                .map(Map::entrySet)
+                .flatMap(Set::stream)
+                .collect(Collectors.groupingBy(
+                        Map.Entry::getKey, Collectors.mapping(Map.Entry::getValue, Collectors.toUnmodifiableList())));
+        return CommitUpdate.invalidateSome(locksTakenOut, locksWithAggregatedMetadata);
```

Comment:
This should be optimal from a complexity perspective and reads nicely, but we might want to use more primitives here in the future

Reply:
outdated, needed to change semantics